### PR TITLE
SI-5082 Cycle avoidance between case companions

### PR DESCRIPTION
**Pull request by retronym**
We can synthesize the case companion unapply without forcing
the info of the case class, by looking at the parameters in
the `ClassDef` tree, rather than at `sym.caseFieldAccessors`.

Review by @paulp. Perhaps you'd like to try it in combination with your `ProductN` patch.

### CODE REVIEW

**Comment by retronym** [comment]
Okay, that was a little optimistic.

```
testing: [...]/files/run/t3569.scala                                  [FAILED]
13d12
< private final int Test$Y.Test$Y$$z2
15c14
< private int Test$Y.Test$Y$$y
---
> private final int Test$Y.z2
16a16
> private int Test$Y.y
```

**Comment by retronym** [comment]
Found a [regression](https://issues.scala-lang.org/browse/SI-7035) en-route...

```
scala> case class Y(final var x: Int, final private var y: String, final val z1: Boolean, final private val z2: Any)
defined class Y

scala> def x = Y.unapply(null)
x: Option[(Int, Boolean, String, Any)]

scala> val Y(a, b, c, d) = Y(0, "", true, new {})
a: Int = 0
b: String = ""
c: Boolean = true
d: Any = $anon$1@56801919

scala> Y.unapply(Y(0, "", true, new {}))
res0: Option[(Int, Boolean, String, Any)] = Some((0,true,,$anon$1@2eac0b4))
```

**Comment by retronym** [comment]
Sadly, I have to withdraw this PR.

To refer the the already-renamed case accessors for non-public fields in the RHS of the unapply method, one cannot work with trees alone.

Here's a monumental hack that actually makes this work, but it is really to inspire us to design our way out of this straight jacket, rather than mergable work: https://github.com/retronym/scala/compare/scala:2.10.x...retronym:ticket/5082-hackety

**Comment by retronym** [comment]
From the ashes...

I think the latest approach is quite workable. Rather than generate the wrong code and patch it up when we can safely peek at the info, we can instead use our privileged seat at the compiler table to convey the accessor renamings through a `perRunMap`.

Another look, @paulp?

**Comment by paulp** [comment]
One pos failure, one neg failure, one run failure. You almost hit for the cycle!

**Comment by retronym** [comment]
`--update-check` is very tempting for this one, we change the secondary error reported after `private[this] invalid for case class field`. But I'll resist for a bit and see if I can't line up all the ducks.

**Comment by adriaanm** [comment]
BUILDLOG?

**Comment by JamesIry** [comment]
LGTM but some stuff I'm not that familiar with. @paulp wanna take a look as well?

File: src/compiler/scala/tools/nsc/typechecker/Unapplies.scala [hunk 1]
```diff
@@ -93,12 +93,33 @@ trait Unapplies extends ast.TreeDSL
...
+      }
+      privateLocalParamAccessor match {
+        case None =>
+          // Selecting by name seems to be the most straight forward way here to
+          // avoid forcing the symbol of the case class in order to list the accessors.
+          val maybeRenamedAccessorName = caseAccessorName(caseclazz.symbol, accessorName)
+          Ident(param) DOT maybeRenamedAccessorName
+        case Some(sym) =>
+          // But, that gives a misleading error message in neg/t1422.scala, where a class
+          // class has an illegal private[this] parameter. We can detect this by checking
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: Just pushed s/class class/case class/

File: src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala [hunk 2]
```diff
@@ -50,6 +50,10 @@ trait SyntheticMethods extends ast.TreeDSL {
     else if (clazz.isDerivedValueClass) valueSymbols
     else Nil
   }
+  private lazy val renamedCaseAccessors = perRunCaches.newMap[Symbol, mutable.Map[TermName, TermName]]()
+  /** Does not force the info of `caseclazz` */
+  final def caseAccessorName(caseclazz: Symbol, paramName: TermName) =
+    (renamedCaseAccessors get caseclazz).fold(paramName)(_(paramName))
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala:56]: It's probably an irrational prejudice but I dislike using placeholder syntax this way. Somehow it's on the other side of my acceptable concision boundary. I prefer

```
(renamedCaseAccessors get caseclazz).fold(paramName)(f => f(paramName))
```

 Just an observation.

Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/SyntheticMethods.scala:56]: Or even

```
(_ apply paramName)
```

**Comment by paulp** [comment]
Looks good to me.

**Pull request merged**